Fix definition of treedef_is_leaf.

pschuh opened the pull request:
Empty nodes like [] and {} have 1 node and 0 leaves. This does not make
them a leaf treedef.

Reproducer:
```
pjit.pjit(lambda x: x, None, (None, {}))((3, {'a': []}))
```

mattjj commented:
This seems right to me, but @hawkinsp can you double check (b/c IIRC you wrote this code when moving pytrees out of Python)?

I think even simpler reproducers are:

```
import jax
print(jax.tree_util.treedef_is_leaf(jax.tree_util.tree_structure({})))
print(jax.tree_util.treedef_is_leaf(jax.tree_util.tree_structure([])))
```

pschuh committed:
Add treedef_is_strict_leaf to fix _prefix_error's semantics.

Empty nodes like [] and {} have 1 node and 0 leaves. This does not make
them a leaf treedef.

Reproducer:
```
pjit.pjit(lambda x: x, None, (None, {}))((3, {'a': []}))
```

pschuh commented:
Ah, I just updated it not to modify the treedef_is_leaf because too many people rely on the broken behavior.
